### PR TITLE
Make nix copy parallel again

### PR DESCRIPTION
**Pull request by thufschmitt**
`nix copy` used to do stuff in parallel. This got removed in https://github.com/NixOS/nix/pull/5048 (to allow batching the copies in the ssh store case), but both can actually be kept in parallel with a little effort.

This is currently incomplete as it
- ~~Disables the batch ssh copy~~
- ~~Removes the ability to copy CA paths from one store prefix to another (I'm curious what the use-case for this is)~~

cc @connorbaker

### CODE REVIEW

**Comment by thufschmitt** [comment]
> This is currently incomplete as it...

Also the code is a bit of a mess right now

File: src/libstore/store-api.cc [hunk 1]
```diff
@@ -998,106 +1078,39 @@ std::map<StorePath, StorePath> copyPaths(
...
+            srcStore.narFromPath(missingPath, sink);
+        });
+        pathsToCopy.push_back(std::pair{*info, std::move(source)});
+    }
 
-    dstStore.addMultipleToStore(*source, repair, checkSigs);
+    dstStore.addMultipleToStore(pathsToCopy, act, repair, checkSigs);
 
     #if 0
-    std::atomic<size_t> nrDone{0};
```
Comment (edolstra) [inline src/libstore/store-api.cc:1024]: You can remove the `#if 0...#endif`.

**Comment by ConnorBaker** [comment]
Thank you so much for doing this @thufschmitt! This change has been invaluable in helping me quickly copy things to a binary cache I set up for CI/CD.

File: src/libstore/store-api.cc (hunk 1 above)
Comment (Ericson2314) [reply src/libstore/store-api.cc:1024]: I am not sure we should remove this. When @edolstra did commented this out, we lost the ability to native `cp` for filesystem -> filesystem stores, and are forced to serialize and deserialize everything.

I think that is bad, and that functionality should be restored.

Overall, want something like "multiple dispatch" for `copyPaths`, where based on both the source and dest stores we can choose more efficient vs more general methods.

File: src/libstore/remote-store.cc [hunk 2]
```diff
@@ -671,6 +671,23 @@ void RemoteStore::addToStore(const ValidPathInfo & info, Source & source,
 }
 
 
+void RemoteStore::addMultipleToStore(
```
Comment (edolstra) [inline src/libstore/remote-store.cc:674]: Shouldn't `LegacySSHStore` do the same thing?

**Comment by edolstra** [comment]
Fixed a subtle hang in f0358ed4650e4608a383bd9f59ee23545f86c4ad.

**Comment by Ericson2314** [comment]
Interesting bug! So this is something that `source`'s own destructor doesn't do? Is that intended?

**Comment by edolstra** [comment]
The problem is that `source`'s destructor doesn't run until the unique_ptr is released.

**Comment by Ericson2314** [comment]
Oh thanks I see now.

I think it might be better to *move* the unique pointer out of the map (leaving behind a null one, I suppose). And then we get the right destructor call by default.

**Comment by edolstra** [comment]
@Ericson2314 You're right, in fact moving it was necessary to get it to build on macOS.

**Pull request merged**